core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java

```diff
@@ -105,79 +123,161 @@ private Deque<Snapshot> orderedChangelogSnapshots(Long fromIdExcl, long toIdIncl
         CloseableIterable<ManifestEntry<DataFile>> entries, TaskContext context) {
 
-      return CloseableIterable.transform(
+      return CloseableIterable.filter(CloseableIterable.transform(
           entries,
           entry -> {
             long commitSnapshotId = entry.snapshotId();
-            int changeOrdinal = snapshotOrdinals.get(commitSnapshotId);
             DataFile dataFile = entry.file().copy(context.shouldKeepStats());
+            DeleteFile[] deleteFiles = context.deletes().forEntry(entry);
 
             switch (entry.status()) {
               case ADDED:
-                return new BaseAddedRowsScanTask(
-                    changeOrdinal,
-                    commitSnapshotId,
-                    dataFile,
-                    NO_DELETES,
-                    context.schemaAsString(),
-                    context.specAsString(),
-                    context.residuals());
-
+                return snapshotOrdinals.containsKey(commitSnapshotId) ?
+                  getAddedRowsScanTask(commitSnapshotId, dataFile, deleteFiles, context) :
+                  getDeletedRowsScanTask(commitSnapshotId, dataFile, deleteFiles, context);
               case DELETED:
-                return new BaseDeletedDataFileScanTask(
-                    changeOrdinal,
-                    commitSnapshotId,
-                    dataFile,
-                    NO_DELETES,
-                    context.schemaAsString(),
-                    context.specAsString(),
-                    context.residuals());
-
+                return getDeletedDataFileScanTask(commitSnapshotId, dataFile, deleteFiles, context);
+              case EXISTING:
+                return getDeletedRowsScanTask(commitSnapshotId, dataFile, deleteFiles, context);
               default:
                 throw new IllegalArgumentException("Unexpected entry status: " + entry.status());
             }
-          });
+          }), Objects::nonNull);
+    }
+
+    private ChangelogScanTask getAddedRowsScanTask(
+        long commitSnapshotId, DataFile dataFile, DeleteFile[] deleteFiles, TaskContext context) {
+      return new BaseAddedRowsScanTask(
+          snapshotOrdinals.get(commitSnapshotId),
+          commitSnapshotId,
+          dataFile,
+          deleteFiles,
+          context.schemaAsString(),
+          context.specAsString(),
+          context.residuals());
+    }
+
+    private ChangelogScanTask getDeletedDataFileScanTask(
+        long commitSnapshotId, DataFile dataFile, DeleteFile[] deleteFiles, TaskContext context) {
+      if (snapshotOrdinals.containsKey(commitSnapshotId)) {
+        return new BaseDeletedDataFileScanTask(
+            snapshotOrdinals.get(commitSnapshotId),
+            commitSnapshotId,
+            dataFile,
+            deleteFiles,
+            context.schemaAsString(),
+            context.specAsString(),
+            context.residuals());
+      } else {
+        // ignore removed data files.
```

Comment:
This case where the commit snapshot id doesn't exist in `snapshotOrdinals` didn't seem to be possible before, what changed?